File: app/src/main/java/com/waygo/data/model/conversation/Waygo.java
package com.waygo.data.model.conversation;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;

import com.waygo.R;
import com.waygo.data.model.butler.ButlerResponse;
import com.waygo.data.model.butler.ButlerSayResponse;
import com.waygo.data.model.butler.ButlerShowResponse;
import com.waygo.utils.option.OptionJ;

public class Waygo extends Person {

    @NonNull
    private final OptionJ<Bitmap> mImage;

    public Waygo(@NonNull final ButlerResponse response) {
        super(response.getMessage(), getUserImage(response));
        mImage = getBitmap(response);
    }

    public Waygo(@NonNull final String message) {
        super(message, OptionJ.asOption(R.drawable.small_waygo));
        mImage = OptionJ.NONE_J;
    }

    @NonNull
    public OptionJ<Bitmap> getImage() {
        return mImage;
    }


    @NonNull
    private static OptionJ<Bitmap> getBitmap(@NonNull final ButlerResponse response) {
        return OptionJ.asOption(response)
                .ofType(ButlerShowResponse.class)
                .map(ButlerShowResponse::getImage);
    }

    @NonNull
    private static OptionJ<Integer> getUserImage(@NonNull final ButlerResponse response) {
        return OptionJ.asOption(response)
                .ofType(ButlerSayResponse.class)
                .map(__ -> R.drawable.small_waygo)
                .orOption(() -> OptionJ.asOption(R.drawable.small_lady));
    }
}

File: app/src/main/java/com/waygo/data/model/conversation/Waygo.kt
package com.waygo.data.model.conversation

import android.graphics.Bitmap

import com.waygo.R
import com.waygo.data.model.butler.ButlerResponse
import com.waygo.data.model.butler.ButlerSayResponse
import com.waygo.data.model.butler.ButlerShowResponse
import com.waygo.utilskt.None
import com.waygo.utilskt.Option
import com.waygo.utilskt.Result

import kotlin.Function0
import kotlin.Function1
import kotlin.platform.platformStatic

public class Waygo : Person {

    public val image: Option<Bitmap>

    public constructor(response: ButlerResponse) : super(response.message, getUserImage(response)) {
        image = getBitmap(response)
    }

    public constructor(message: String) : super(message, Option.ofObj(R.drawable.small_waygo)) {
        image = None<Bitmap>()
    }

    private fun getBitmap(response: ButlerResponse): Option<Bitmap> {
        return Option.ofObj(response)
                .ofType(javaClass<ButlerShowResponse>())
                .map{it.image}
    }

    companion object {
        private platformStatic fun getUserImage(response: ButlerResponse): Option<Int> {
            return Option.ofObj(response)
                    .ofType(javaClass<ButlerSayResponse>())
                    .map {R.drawable.small_waygo}
                    .orOption {Option.ofObj(R.drawable.small_lady)}
        }
    }
}
